[Rollup] Add `wait_for_completion` option to StopRollupJob API

polyfractal opened the pull request:
This adds a flag which allows the user to block the Stop API until the task has actually moved to a stopped state, instead of returning immediately.

If the flag is unset (default) the behavior is as before: returns immediately and stopping happens async in the background.  If the flag is set to a TimeValue, it will block the call until the status changes to `STOPPED` (up to the specified time limit). If the time value is exceeded a timeout exception is thrown.

**Question:** The current code executes the Tasks operation on the `SAME` thread, which I'm assuming would be the networking thread.  Since this opens the chance of blocking the thread temporarily, should we switch to `MANAGEMENT` or `GENERIC` or something?

Also, this is technically an enhancement but critical to a Kibana blocker, so I'd lobby that it be allowed into 6.5

Followup to #34574

/cc @jen-huang @cjcenizal

elasticmachine commented:
Pinging @elastic/es-search-aggs

polyfractal commented:
Quick note: if we find this too risky to put in 6.5, #34574 should largely negate the race condition that Kibana was worried about.  If this PR doesn't merge the behavior will just be: 

user hits stop, spinny for a second, hits delete, gets an error about job still running, tries again (`n` times) until it finally deletes.

Not wonderful, but not end of the world either.

jimczi commented on docs/reference/rollup/apis/stop-job.asciidoc:
```diff
@@ -22,6 +22,14 @@ Stopping an already stopped job has no action.
 
+==== Query Parameters
+
+`wait_for_stopped` (optional)::
```
`wait_for_completion` for consistency with the other APIs that waits for the completion of the command ?

jimczi replied:
Also it might be difficult for a user to set up a value for the timeout. So maybe we should make it a `boolean` and have another parameter (`timeout`)  that could default to `30s` ?

polyfractal replied:
Yeah, I was hoping to avoid two parameters, since configuring a timeout without setting `wait_for_completion` to true won't do anything.  But I can see how it's a strange setup.  I could go either way

jimczi replied:
The two params solution is consistent with the GetTask API but I don't have a strong opinion here either.

jimczi commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportStopRollupAction.java:
```diff
@@ -52,12 +57,52 @@ protected void taskOperation(StopRollupJobAction.Request request,
                                  ActionListener<StopRollupJobAction.Response> listener) {
         if (jobTask.getConfig().getId().equals(request.getId())) {
             jobTask.stop(listener);
```
we should not call the `listener` if `waitForStopped` is set.

polyfractal replied:
urgh yes, that would have been bad...

jimczi commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportStopRollupAction.java:
```diff
@@ -52,12 +57,52 @@ protected void taskOperation(StopRollupJobAction.Request request,
                                  ActionListener<StopRollupJobAction.Response> listener) {
         if (jobTask.getConfig().getId().equals(request.getId())) {
             jobTask.stop(listener);
+            waitForStopped(request, jobTask, listener);
```
We shouldn't block the current thread. I think we can use the `generic` thread pool here in order to make sure that we don't block on important threads (network threads). Something like:
````
 threadPool.generic().execute(new AbstractRunnable() {
                   @Override
                    protected void doRun() throws Exception {
                          if (waitForStopped(request, jobTask)) {
                            listener.onResponse(new Stop...);
                          } else {
                            listener.onFailure(new Stop...);
                         }
                    }

                    @Override
                    public void onFailure(Exception e) {
                        listener.onFailure(e);
                    }

polyfractal replied:
Should we just have the task run on the Generic thread from the start?  E.g. in the ctor just specify `generic` instead of `same`?

jimczi replied:
I prefer if we explicitly fork to another thread with a comment explaining why we do that.  But yes I think it would work if we use `generic` directly. No strong opinion here either, just a preference ;)

polyfractal replied:
++

polyfractal commented:
@jimczi pending CI passing again I think I addressed your comments

- swapped to `wait_for_completion`, added a `timeout`.  If `timeout` is omitted the default is 30s like the other task APIs.  If `wait_for_completion: false` then the timeout doesn't do anything
- Spawns to a generic threadpool on demand, as discussed.
- Refactored so that the listener is wrapped correctly instead of being called twice.

polyfractal commented:
Right, trying this again.  This time it has the generic threadpool stuff :)

polyfractal commented:
Jenkins, run gradle build tests please
